### PR TITLE
Add tests for invoice finance flow

### PR DESCRIPTION
**Pull request by sandeepsajan0**
Fixes #ISSUEID

Add description here.

### CODE REVIEW

File: hypha/apply/api/v1/tests/test_serializers.py [hunk 1]
```diff
@@ -26,3 +31,45 @@ def test_handles_negative_reviews(self):
         self.assertEqual(data['recommendation'], {'value': 0, 'display': 'No'})
         self.assertEqual(len(data['assigned']), 1)
         self.assertEqual(len(data['reviews']), 1)
+
+
+class TestDeliverableSerializer(TestCase):
+    def test_id_is_required(self):
+        serializer = DeliverableSerializer(data={'quantity': 1})
+        self.assertFalse(serializer.is_valid(), serializer.error_messages)
```
Comment (sks444) [inline hypha/apply/api/v1/tests/test_serializers.py]: You can also add another assert, assertEqual, to actually validate the error message that is being raised.

File: hypha/apply/api/v1/tests/test_serializers.py [hunk 2]
```diff
@@ -26,3 +31,45 @@ def test_handles_negative_reviews(self):
...
+    def test_quantity_not_required(self):
+        deliverable = DeliverableFactory()
+        serializer = DeliverableSerializer(data={'id': deliverable.id})
+        self.assertTrue(serializer.is_valid())
+
+
+class TestInvoiceRequiredChecksSerializer(TestCase):
+    def test_valid_checks_required(self):
+        serializer = InvoiceRequiredChecksSerializer(data={'valid_checks_link': 'https://google.com'})
+        self.assertFalse(serializer.is_valid())
```
Comment (sks444) [inline hypha/apply/api/v1/tests/test_serializers.py:63]: Also validate the error message string that is being raised.

File: hypha/apply/api/v1/tests/test_views.py [hunk 3]
```diff
@@ -90,3 +107,242 @@ def test_out_of_order_does_nothing(self):
...
+
+    def test_deliverable_dont_exists_in_project_deliverables(self):
+        user = StaffFactory()
+        project = ProjectFactory()
+        invoice = InvoiceFactory(project=project, status=SUBMITTED)
+        deliverable = DeliverableFactory()
+        self.client.force_login(user)
+
+        response = self.post_to_add(project.id, invoice.id, deliverable.id)
+        self.assertEqual(response.status_code, 400)
```
Comment (sks444) [inline hypha/apply/api/v1/tests/test_views.py:277]: Also validate the error message that is being raised.

File: hypha/apply/api/v1/tests/test_views.py [hunk 4]
```diff
@@ -90,3 +107,242 @@ def test_out_of_order_does_nothing(self):
...
+        user = StaffFactory()
+        project = ProjectFactory()
+        invoice = InvoiceFactory(project=project, status=SUBMITTED)
+        deliverable = DeliverableFactory(project=project)
+        invoice_deliverable = InvoiceDeliverableFactory(deliverable=deliverable)
+        invoice.deliverables.add(invoice_deliverable)
+        self.client.force_login(user)
+
+        response = self.post_to_add(project.id, invoice.id, deliverable.id)
+        self.assertEqual(response.status_code, 400)
```
Comment (sks444) [inline hypha/apply/api/v1/tests/test_views.py:291]: Also validate the error message that is being raised.

File: hypha/apply/api/v1/tests/test_views.py [hunk 5]
```diff
@@ -90,3 +107,242 @@ def test_out_of_order_does_nothing(self):
...
+
+    def test_deliverable_available_gte_quantity(self):
+        user = StaffFactory()
+        project = ProjectFactory()
+        invoice = InvoiceFactory(project=project, status=SUBMITTED)
+        deliverable = DeliverableFactory(project=project)
+        self.client.force_login(user)
+
+        response = self.post_to_add(project.id, invoice.id, deliverable.id, quantity=3)
+        self.assertEqual(response.status_code, 400)
```
Comment (sks444) [inline hypha/apply/api/v1/tests/test_views.py:303]: Also validate the error message that is being raised.

**Review by sks444 — COMMENTED** [review]
Looking good @sandeepsajan0, nice progress!

File: hypha/apply/api/v1/tests/test_views.py [hunk 6]
```diff
@@ -90,3 +108,249 @@ def test_out_of_order_does_nothing(self):
         self.assertEqual(response_one.status_code, 200, response_one.json())
         self.assertEqual(response_two.status_code, 404, response_two.json())
         self.assertEqual(Activity.objects.count(), 2)
+
+
+@override_settings(ROOT_URLCONF='hypha.apply.urls')
+class TestInvoiceDeliverableViewset(TestCase):
```
Comment (sks444) [inline hypha/apply/api/v1/tests/test_views.py:115]: @sandeepsajan0 can we also add a test with not letting deliverables to be removed once the invoice has been approved.

Staff can not removed deliverables if the invoice is in APPROVED BY STAFF state
Finance 1/Staff can not remove deliverables if the invoice is in APPROVED BY FINANCE 1 state
No one can remove deliverables once the invoice has been approved by finance 2

Basically a few tests for removing deliverables based on user and invoice state.

Comment (sandeepsajan0) [reply hypha/apply/api/v1/tests/test_views.py:115]: Yes, Sure. That would make it more covered.

**Review by sks444 — APPROVED** [review]
Looks good. @frjo we can merge this in.

**Pull request merged**